tools/functional-tester/etcd-runner/main.go

```diff
@@ -212,6 +215,170 @@ func runRacer(eps []string, round int) {
+	reqRate := 20                      // put request per second
+	maxKeys := 100000                  // max number of keys for put operation
+
+	ctx := context.Background()
+	prefixKeysMap := make(map[string][]string) // map for holding prefixes and values
+	ws := make([]watcher, noOfPrefixes*watchPerPrefix)
+	// pre generate prefixes equal to noOfPrefixes
+	prefixes := generateUniqueKeys(5, noOfPrefixes)
+	// pre generate keys
+	keys := generateRandomKeys(10, maxKeys)
+
+	var (
+		wg  sync.WaitGroup
+		pr  *clientv3.PutResponse
+		err error
+	)
+
+	// assign watchPerPrefix times watchers on each prefix
+	wg.Add(len(ws))
+	index := 0
+	for i := 0; i < len(ws); {
+		prefix := prefixes[index]
+		// initialize prefix key map for each prefix
+		prefixKeysMap[prefix] = make([]string, 0)
+		for j := 0; j < watchPerPrefix; j++ {
+			go func(w *watcher, prefix string) {
+				defer wg.Done()
+				w.c = randClient(eps)
+				w.prefix = prefix
+				w.wc = w.c.Watch(ctx, w.prefix, clientv3.WithPrefix())
+			}(&ws[i], prefix)
+			i++
+		}
+		index++
+	}
+	wg.Wait()
+
+	ctxt, cancel := context.WithDeadline(ctx, time.Now().Add(runningTime*time.Second))
+	defer cancel()
+	// generate and put keys in cluster
+	go func() {
+		client := randClient(eps)
+		defer client.Close()
+
+		limiter := rate.NewLimiter(rate.Limit(reqRate), reqRate)
+
+		for i := 0; i < len(keys); i++ {
+			randPrefixIndex := rand.Intn(len(prefixes))
+			prefix := prefixes[randPrefixIndex]
+			key := prefix + keys[i]
+			// limit key put as per reqRate
+			err = limiter.Wait(ctxt)
+			if err != nil {
+				break
+			}
+
+			//add all possible updates on prefixes for the generated key
+			for prefixkey, vals := range prefixKeysMap {
+				if strings.HasPrefix(key, prefixkey) {
+					prefixKeysMap[prefixkey] = append(vals, key)
+				}
+			}
+
+			// perform the put operation
+			pr, err = client.Put(ctxt, key, key)
+			if err == context.DeadlineExceeded {
+				break
+			}
+
+			if err != nil {
+				log.Printf("Error: %v occured while trying to key: %v, value : %v to kv store.", err, key, key)
+				continue
+			}
+		}
+	}()
+
+	wg.Add(len(ws))
+	for i := range ws {
+		// for all watcher verify the wtch out put with actual key updates on prefixes
+		go func(w *watcher) {
+			defer wg.Done()
+			index := 0
+			for {
+				select {
+				case watchChan := <-w.wc:
+					keys := prefixKeysMap[w.prefix]
+					retrievedVal := string(watchChan.Events[0].Kv.Value)
+					val := keys[index]
+					if val != retrievedVal {
+						log.Fatal("expected value %q, got %q for prefix : %q\n", val, retrievedVal, w.prefix)
+					}
+					index++
+				case <-ctxt.Done():
+					if ctxt.Err() == context.Canceled || ctxt.Err() == context.DeadlineExceeded {
+						return
+					}
+				default:
+				}
+			}
+		}(&ws[i])
+	}
+	wg.Wait()
+
+	// close all clients
+	for _, w := range ws {
+		w.c.Close()
+	}
+}
+
+func generateUniqueKeys(maxstrlen int, keynos int) []string {
+	prefixMap := make(map[string]bool)
```

Comment:
This is variable is not being referred anywhere in our main logic. Will change the name to look generic.